Adding helper function to SSH into a Kubevirt VM

mkoppal-px opened the pull request:
**What this PR does / why we need it**:

- [x] Added helper function `RunCmdInVM` to SSH into a VM and run a command
- [x] Added label `username` to VM spec from where the username for login will be fetched
- [x] Added openssh and sshpass utility to torpedo image
- [x] Added a new kubevirt application spec - `kubevirt-cirros-container-disk`
- [x] Added `VirtualMachineInstanceOps` interface to sched-ops and vendored it in

**Which issue(s) this PR fixes** (optional)
Closes #PB-4793

**Special notes for your reviewer**:

- The `KubevirtTest` is a dummy test to verify if SSH works correctly
- [Jenkins run on Vanilla](https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/4402)
- [Jenkins run on OCP](https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/4376/)
- [Jenkins run with SSH on VM in destination cluster](https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/4373/)

vsundarraj-px commented on drivers/scheduler/k8s/specs/kubevirt-vm-pvc/service.yaml:
```diff
@@ -0,0 +1,27 @@
+    app: kubevirt-vm-pvc
+  type: NodePort
+
+#apiVersion: v1
```
Need to remove commented items

mkoppal-px replied:
Done

dbhatnagar-px commented on tests/backup/backup_kubevirt_test.go:
```diff
@@ -669,3 +669,47 @@ var _ = Describe("{KubevirtUpgradeTest}", func() {
 	})
 })
+
+var _ = Describe("{KubevirtTest}", func() {
```
Can we rename this to KubevirtVMSshTest to provide more clarity

mkoppal-px replied:
Done

dbhatnagar-px commented on tests/backup_helper.go:
```diff
@@ -5914,6 +5915,166 @@ func RestartAllVMsInNamespace(namespace string, waitForCompletion bool) error {
+func initSSHPod() error {
+	var p *corev1.Pod
+	var err error
+	if p, err = k8sCore.GetPodByName(sshPodName, "default"); p == nil {
```
Let's not run it in default ns but either let's run this pod in same namespace as VM we want to ssh to or we can create a different namespace for this pod to run in altogether

mkoppal-px replied:
The reason I did this was to avoid spawning multiple pods in multiple namespaces. Now if I have to run ssh commands in 10 VMs which are in 10 different namespaces, then 1 SSH pod in default namespace would suffice. As opposed to created 10 SSH pods.
But yes, I can switch from default to some other dedicated namespace.

mkoppal-px replied:
Creating a ns called `ssh-pod-namespace` and bring up the pod there.

dbhatnagar-px commented on tests/backup_helper.go:
```diff
@@ -5914,6 +5915,166 @@ func RestartAllVMsInNamespace(namespace string, waitForCompletion bool) error {
+			return err
+		}
+		testCmdArgs := []string{"sshpass", "-p", password, "ssh", "-o", "StrictHostKeyChecking=no", fmt.Sprintf("cirros@%s", ipAddress), "hostname"}
+		cmdArgs := []string{"sshpass", "-p", password, "ssh", "-o", "StrictHostKeyChecking=no", fmt.Sprintf("cirros@%s", ipAddress), cmd}
```
We can avoid duplicacy in these 2 vars here

mkoppal-px replied:
Create a func like this 
```
func getSSHCommandArgs(username, password, ipAddress, cmd string) []string {
	return []string{"sshpass", "-p", password, "ssh", "-o", "StrictHostKeyChecking=no", fmt.Sprintf("%s@%s", username, ipAddress), cmd}
}
```

And the using the variables as follows:
```
testCmdArgs := getSSHCommandArgs(username, password, ipAddress, "hostname")
		cmdArgs := getSSHCommandArgs(username, password, ipAddress, cmd)
```
I need `testCmdArgs` to check if the connection is going through and run a sample command.

dbhatnagar-px reviewed (approved):
Lgtm

kshithijiyer-px commented on tests/backup_helper.go:
```diff
@@ -5914,6 +5916,187 @@ func RestartAllVMsInNamespace(namespace string, waitForCompletion bool) error {
 	return nil
 }
 
+func GetAllVMsInNamespace(namespace string) ([]kubevirtv1.VirtualMachine, error) {
```
Can we add a doc string here?

mkoppal-px replied:
Done

kphalgun-px commented on tests/backup/backup_kubevirt_test.go:
```diff
@@ -669,3 +669,47 @@ var _ = Describe("{KubevirtUpgradeTest}", func() {
+			}
+			for _, vm := range vms {
+				log.Infof("Running command for VM [%s]", vm.Name)
+				err = RunCmdInVM(vm, "uname -a", ctx)
```
How about?

```go
import "context"

err = RunCmdInVM(vm, "uname -a", context.Background())
```

mkoppal-px replied:
This is just a dummy test case. We can pass `context.Background()` when we write the real test case.

kphalgun-px commented on tests/backup_helper.go:
```diff
@@ -5914,6 +5916,187 @@ func RestartAllVMsInNamespace(namespace string, waitForCompletion bool) error {
+}
+
+// RunCmdInVM runs a command in the VM by SSHing into it
+func RunCmdInVM(vm kubevirtv1.VirtualMachine, cmd string, ctx context1.Context) error {
```
How about we return output as well?

mkoppal-px replied:
What output are you expecting here?

kphalgun-px replied:
output from command?

mkoppal-px replied:
Ok, will do it. You are right, in some cases we may need to check something in the output string.

kphalgun-px commented on tests/backup_helper.go:
```diff
@@ -5914,6 +5916,187 @@ func RestartAllVMsInNamespace(namespace string, waitForCompletion bool) error {
+	log.Infof("SSH Command - %s", sshCmd)
+
+	// If the cluster provider is openshift then we are creating ssh pod and running the command in it
+	if os.Getenv("CLUSTER_PROVIDER") == "openshift" {
```
How about?

```go
import "github.com/portworx/torpedo/drivers/scheduler/openshift"

Inst().S.String() == openshift.SchedName
```

mkoppal-px replied:
I thought about this, but today we are passing scheduler as k8s even for OCP like this `"--scheduler", "k8s"`
Once we start passing `openshift` for `scheduler`. I will change it.

mkoppal-px commented:
Latest run after addressing review comments - https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/4467

kshithijiyer-px commented on tests/backup/backup_kubevirt_test.go:
```diff
@@ -669,3 +669,48 @@ var _ = Describe("{KubevirtUpgradeTest}", func() {
+
+		Step("SSH into the kubevirt VM", func() {
+			log.Infof("Sleeping...")
+			//time.Sleep(1 * time.Minute)
```
Let's get rid of this and the corresponding log message as we aren't using it.

kshithijiyer-px reviewed (approved):
Minor nit fix rest looks good to me.

kphalgun-px commented on tests/backup_helper.go:
```diff
@@ -5985,6 +5987,188 @@ func RestartAllVMsInNamespace(namespace string, waitForCompletion bool) error {
+		cmdArgs := getSSHCommandArgs(username, password, ipAddress, cmd)
+
+		// To check if the ssh server is up and running
+		t := func() (interface{}, bool, error) {
+			output, err := k8sCore.RunCommandInPod(testCmdArgs, sshPodName, "ssh-container", "default")
+			if err != nil {
+				log.Infof("Error encountered")
+				if isConnectionError(err.Error()) {
+					log.Infof("Test connection output - \n%s", output)
+					return "", true, err
+				} else {
+					return "", false, err
+				}
+			}
+			log.Infof("Test connection success output - \n%s", output)
+			return "", false, nil
+		}
+		_, err = task.DoRetryWithTimeout(t, 10*time.Minute, 30*time.Second)
+		if err != nil {
+			return "", err
+		}
```
How about?

```go
		t := func() (interface{}, bool, error) {
			output, err := k8sCore.RunCommandInPod(testCmdArgs, sshPodName, "ssh-container", "default")
			if err != nil {
				log.Infof("Error encountered")
				if isConnectionError(err.Error()) {
					log.Infof("Test connection output - \n%s", output)
					return "", true, err
				} else {
					return output, false, err
				}
			}
			log.Infof("Test connection success output - \n%s", output)
			return output, false, nil
		}
		output, err = task.DoRetryWithTimeout(t, 10*time.Minute, 30*time.Second)
		if err != nil {
			return "", err
		}
		return output, nil
```

mkoppal-px commented on tests/backup_helper.go:
```diff
@@ -5985,6 +5987,188 @@ func RestartAllVMsInNamespace(namespace string, waitForCompletion bool) error {
+					log.Infof("Output of cmd %s - \n%s", cmd, output)
+					return "", true, err
+				} else {
+					return "", false, err
```
Done
